### ISSUE TITLE
[kubernetes] Race condition with K8s Operator and CRD

### ISSUE THREAD

**Issue by tgaddair** [ray-project/ray#15744]
### What is the problem?

*Ray version and other system information (Python version, TensorFlow version, OS):* 1.3.0

There appears to be a race condition between the Ray Operator pod and the RayCluster CRD in cases where they are deployed together (e.g., Helm). Namely, it seems that if the Ray Operator starts first and hits [the line](https://github.com/ray-project/ray/blob/4459a5a2b1f1ba1376f5773c02916df01e47c403/python/ray/ray_operator/operator_utils.py#L80) where it begins to watch for events before the CRD is up, then no events will ever be picked up.

### Reproduction (REQUIRED)

Use Helm to deploy both the CRD and the Ray Operator together, or perhaps try deploying the operator first (though it is possible the specific issue comes from the CRD being partially initialized, not sure).

Then try creating a Ray cluster and note that no logs are ever emitted from the Ray Operator. Then restart the operator. Suddenly it can detect and properly allocate RayCluster pods.

cc @DmitriGekhtman

**Comment by DmitriGekhtman** [comment]
It's possible the CRD is partially initialized.
Could you give more details about how you're currently deploying?

**Comment by DmitriGekhtman** [comment]
I _think_ Helm should do the right thing if the crd is stuck under `crds`.
I'll make a note in the doc update that's coming with the Helm PR that, if deploying manually, one should deploy the CRD first and give it a few seconds to register.

**Comment by tgaddair** [comment]
Earlier today I noticed a few other occurrences of this problem (Ray Clusters not being detected / initialized correctly). This was for a Ray Operator that was already running. In general, my suspicion here is that responding to k8s events is not going to be 100% reliable.

1. In one case, the added RayCluster failed to get picked up after a monitoring process in the Ray Operator crashed the the following exception:

```
monitor.py:253 -- Error in monitor loop
Traceback (most recent call last):
  File "/home/ray/anaconda3/lib/python3.7/site-packages/ray/_private/monitor.py", line 284, in run
    self._run()
  File "/home/ray/anaconda3/lib/python3.7/site-packages/ray/_private/monitor.py", line 175, in _run
    self.update_load_metrics()
  File "/home/ray/anaconda3/lib/python3.7/site-packages/ray/_private/monitor.py", line 140, in update_load_metrics
    request, timeout=4)
  File "/home/ray/anaconda3/lib/python3.7/site-packages/grpc/_channel.py", line 946, in __call__
    return _end_unary_response_blocking(state, call, False, None)
  File "/home/ray/anaconda3/lib/python3.7/site-packages/grpc/_channel.py", line 849, in _end_unary_response_blocking
    raise _InactiveRpcError(state)
grpc._channel._InactiveRpcError: <_InactiveRpcError of RPC that terminated with:
	status = StatusCode.UNAVAILABLE
	details = "Socket closed"
	debug_error_string = "{"created":"@1620849572.209657346","description":"Error received from peer ipv4:10.244.5.5:36945","file":"src/core/lib/surface/call.cc","file_line":1067,"grpc_message":"Socket closed","grpc_status":14}"

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/ray/anaconda3/lib/python3.7/multiprocessing/process.py", line 297, in _bootstrap
    self.run()
  File "/home/ray/anaconda3/lib/python3.7/multiprocessing/process.py", line 99, in run
    self._target(*self._args, **self._kwargs)
  File "/home/ray/anaconda3/lib/python3.7/site-packages/ray/ray_operator/operator.py", line 54, in _create_or_update
    self.start_monitor()
  File "/home/ray/anaconda3/lib/python3.7/site-packages/ray/ray_operator/operator.py", line 79, in start_monitor
    self.mtr.run()
  File "/home/ray/anaconda3/lib/python3.7/site-packages/ray/_private/monitor.py", line 286, in run
    self._handle_failure(traceback.format_exc())
  File "/home/ray/anaconda3/lib/python3.7/site-packages/ray/_private/monitor.py", line 264, in _handle_failure
    _internal_kv_put(DEBUG_AUTOSCALING_ERROR, message, overwrite=True)
  File "/home/ray/anaconda3/lib/python3.7/site-packages/ray/_private/client_mode_hook.py", line 47, in wrapper
    return func(*args, **kwargs)
  File "/home/ray/anaconda3/lib/python3.7/site-packages/ray/experimental/internal_kv.py", line 57, in _internal_kv_put
    key, "value", value)
  File "/home/ray/anaconda3/lib/python3.7/site-packages/redis/client.py", line 3050, in hset
    return self.execute_command('HSET', name, *items)
  File "/home/ray/anaconda3/lib/python3.7/site-packages/redis/client.py", line 898, in execute_command
    conn = self.connection or pool.get_connection(command_name, **options)
  File "/home/ray/anaconda3/lib/python3.7/site-packages/redis/connection.py", line 1202, in get_connection
connection.connect()
  File "/home/ray/anaconda3/lib/python3.7/site-packages/redis/connection.py", line 563, in connect
    raise ConnectionError(self._error_message(e))
redis.exceptions.ConnectionError: Error 113 connecting to 10.244.5.5:6379. No route to host.
```

2. After restarting the Ray Operator, it still failed to initialize the job, as it kept trying to login to the pod, but never created the pod (for some reason):

```
default-cluster:2021-05-12 13:10:19,833	INFO command_runner.py:172 -- NodeUpdater: default-cluster-ray-head-2ggf9: Running kubectl -n predibase exec -it default-cluster-ray-head-2ggf9 -- bash --login -c -i 'true && source ~/.bashrc && export OMP_NUM_THREADS=1 PYTHONWARNINGS=ignore && (uptime)'
Unable to use a TTY - input is not a terminal or the right kind of file
Error from server (BadRequest): pod default-cluster-ray-head-2ggf9 does not have a host assigned
```

All of this to say that I think the root of the issue here is that the Ray Operator is doing work that is stateful without a persistent store, which is leading to things getting into invalid states from which it can't recover gracefully.

My suggestion (at a very high level) is that we may want to rethink some of the architecture of the Ray Operator in one of two directions:

1. Introduce a persistent storage layer, like Zookeeper or any DB, to keep track of cluster lifecycles and other state in case of pod restarts, etc.
2. Rearchitect operator components to be as stateless as possible and rely on more self-initialization of the head node, workers, and autoscaler, along with active polling from the operator to health / liveness check the ray cluster.

As a short term solution to this problem, one option might be to add a background thread in the operator that exists alongside the event stream watcher that polls for all RayClusters and check their status. If it detects that the cluster is not running, for example, it could initiate the add cluster workflow [here](https://github.com/ray-project/ray/blob/6e3ec98b729b56754a62de68d9c4029123795b43/python/ray/ray_operator/operator.py#L245).

What do you think?

**Comment by DmitriGekhtman** [comment]
I think the particular issues here have been fixed in current master.

It's kind of a janky setup, but each monitor update queries the GCS, effectively serving as a health check on the Ray Cluster.
https://github.com/ray-project/ray/blob/cdf93930f3349a03e68d844ccbef940c5b84c68e/python/ray/autoscaler/_private/monitor.py#L184-L186

With the recent changes, a Ray failure does trigger a Ray restart via a "MODIFIED" event.

That said, we could definitely expand state tracking, etc.

I think the real gap Ray/Kubernetes gap is that the Ray GCS, which is at the immediate moment not fault-tolerant at all lives inside of a Pod, which is ephemeral pretty much by definition.
Basically, the entire operational state of the Ray cluster is encased in gasoline-soaked papier-mâché. (Ok maybe this is a little extreme.)

IIUC correctly fault tolerance / backend-pluggability for the GCS is currently ongoing work of the Ray Core? cc @ericl

**Comment by tgaddair** [comment]
One difference here is that the Ray cluster itself is being started for the first time, so it's a failure of the Ray Operator to detect a CREATE event, which may be different from what you're referring to.

**Comment by DmitriGekhtman** [comment]
Hmm, that is strange - this could use repro with master, though. (SHA can be pinned, nightly operator compatible with 1.3.0 cluster.)
A bunch of fixes have been applied since 1.3.0 branch cut in March.

**Comment by tgaddair** [comment]
I'm going to be running some tests today, but truthfully I can see by looking at the code that there are a number of stateful elements of the existing operator that will cause these kinds of failures in the event of pod restart. For example, if you look at the deletion logic [here](https://github.com/ray-project/ray/blob/master/python/ray/ray_operator/operator.py#L300), it seems that if the pod restarts, this dict will be empty and the deletion logic will fail. Similarly for the configs stored locally within the container. So these things need to be addressed for this to be stable as a long running deployment.

**Comment by DmitriGekhtman** [comment]
Hmm, it is currently not well commented, but if the operator pod restarts, the operator will register harmless ADDED events for all existing RayCluster resources. This will run the equivalent of ``ray up --no-restart``  for all of the running clusters, repopulate the dict, and re-store the configs, without restarting the Ray cluster.

The lack of proper external state management will probably cause problems somewhere else, but I think in this case the behavior is more or less as expected.

**Comment by DmitriGekhtman** [comment]
Also, the configs should instead be stored as ConfigMaps..
